### DISCUSSION TITLE
Start in incognito mode

### DISCUSSION THREAD

**Discussion by zerowebcorp** [php-webdriver/php-webdriver#793] [Q&A] [answered]
Hello

I am using selenium hub and chrome in a docker container (https://github.com/SeleniumHQ/docker-selenium) 

When using the webdriver, I am not able to start an incognito window.

```
$options = new ChromeOptions();
$capabilities = DesiredCapabilities::chrome();
$capabilities->setCapability('acceptInsecureCerts', true);
$capabilities->setCapability('chromeOptions', $options);
$capabilities->setCapability('goog:chromeOptions', ['args' => ['--incognito']]);
$driver = RemoteWebDriver::create('http://selenium-hub:4444/wd/hub', $capabilities);
```

I have also tried with 
```
$options->addArguments([
            '--incognito',
        ]);
```

Output

```
 #capabilities: Facebook\WebDriver\Remote\DesiredCapabilities^ {#130
    -capabilities: array:19 [
      "acceptInsecureCerts" => true
      "browserName" => "chrome"
      "browserVersion" => "81.0.4044.92"
      "chrome" => array:2 [
        "chromedriverVersion" => "81.0.4044.69 (6813546031a4bc83f717a2ef7cd4ac6ec1199132-refs/branch-heads/4044@{#776})"
        "userDataDir" => "/tmp/.com.google.Chrome.Oqy8DP"
      ]
      "goog:chromeOptions" => array:1 [
        "debuggerAddress" => "localhost:34571"
      ]
      "networkConnectionEnabled" => false
      "pageLoadStrategy" => "normal"
      "platformName" => "linux"
      "proxy" => []
      "setWindowRect" => true
      "strictFileInteractability" => false
      "timeouts" => array:3 [
        "implicit" => 0
        "pageLoad" => 300000
        "script" => 30000
      ]
      "unhandledPromptBehavior" => "dismiss and notify"
      "webauthn:virtualAuthenticators" => true
      "webdriver.remote.sessionid" => "aeea5ee480d264bf4520ba208f492b8a"
      "platform" => "linux"
      "version" => "81.0.4044.92"
      "acceptSslCerts" => true
      "chromeOptions" => array:1 [
        "debuggerAddress" => "localhost:34571"
      ]
    ]
  }
```

I am not finding any documentation around opening an incognito window. Is this supported? 

php-webdriver/webdriver version 1.8.2
Selenium hub 3.141.59-20200409
Chrome Node node-chrome-debug:3.141.59-20200409
Chrome Drier 81.0.4044.69

**Comment by williamdes** [comment]
Could you try: https://stackoverflow.com/a/19026738/5155484

**Comment by zerowebcorp** [comment]
@williamdes  Thanks for responding. I did try this a few minutes ago. No luck

```
 $capabilities->setCapability('chrome.switches', ['--incognito']);
```
Input
![image](https://user-images.githubusercontent.com/914305/82346075-084b1700-99c4-11ea-91cc-7de0a17d4776.png)

Output


![image](https://user-images.githubusercontent.com/914305/82345949-dafe6900-99c3-11ea-82a8-bab7fd27705a.png)

**Comment by OndraM** [comment]
What is the point of starting browser via Selenium in incognito mode?

When using WebDriver automation, browser creates new (empty) session when started, and destroys it when done. Which is almost the same as incognito mode.

**Comment by OndraM** [comment, marked as answer]
This seems to be working for me:

```php
$chromeOptions = new ChromeOptions();
$chromeOptions->addArguments(['--incognito']);

$capabilities = DesiredCapabilities::chrome();
$capabilities->setCapability('chromeOptions', $chromeOptions);
```

The browser is then dark and says "Incognito mode". However, I still think this has no real point, besides black color of the browser :upside_down_face: .

**Comment by zerowebcorp** [comment]
That was my understanding. But what I am finding is that when we run two same scripts in parallel, the first session creates some cookies which the second session is able to detect. Because of that cookie the second session fails. Hence I need to open an incognito

**Comment by OndraM** [comment]
> That was my understanding. But what I am finding is that when we run two same scripts in parallel, the first session creates some cookies which the second session is able to detect. Because of that cookie the second session fails.

How could this happen? If you create new session for each scenario, their cookies will be isolated and cannot be read by another instance... :thinking: .

Do you have any reproducable scenario how cookies created in one session could be read in another session? How do you start WebDriver? Don't you reuse browser session?

**Comment by zerowebcorp** [comment]
```
     $capabilities = DesiredCapabilities::chrome();
        $capabilities->setCapability('acceptInsecureCerts', true);
        $capabilities->setCapability('goog:chromeOptions', [
            'args' => [
                '--incognito',
            ]
        ]);
```

This now works for me. I can see the dark chrome window.

**Comment by williamdes** [comment]
A TLDR comment:

> What is the point of starting browser via Selenium in incognito mode?

Some browsers have "stupid" behaviors like blocking scripts (once Mozilla did block the Facebook login script in private mode and this did crash the Angular website behind. :) )
